From fa45ae0037bedc9969dbfefcd1da07a888b4a30a Mon Sep 17 00:00:00 2001
From: Dexuan Cui <decui@microsoft.com>
Date: Mon, 14 Dec 2015 16:01:47 -0800
Subject: Drivers: hv: vmbus: serialize process_chn_event() and
 vmbus_close_internal()
Reply-To: joseph.salisbury@microsoft.com

process_chn_event(), running in the tasklet, can race with
vmbus_close_internal() in the case of SMP guest, e.g., when the former is
accessing channel->inbound.ring_buffer, the latter could be freeing the
ring_buffer pages.

To resolve the race, we can serialize them by disabling the tasklet when
the latter is running here.

Signed-off-by: Dexuan Cui <decui@microsoft.com>
Signed-off-by: K. Y. Srinivasan <kys@microsoft.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
(backported from commit 63d55b2aeb5e4faa170316fee73c3c47ea9268c7)
Signed-off-by: Joseph Salisbury <joseph.salisbury@microsoft.com>
---
 drivers/hv/channel.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/hv/channel.c b/drivers/hv/channel.c
index d037454fe7b8..73263095e475 100644
--- a/drivers/hv/channel.c
+++ b/drivers/hv/channel.c
@@ -565,6 +565,12 @@ static int vmbus_close_internal(struct vmbus_channel *channel)
 	free_pages((unsigned long)channel->ringbuffer_pages,
 		get_order(channel->ringbuffer_pagecount * PAGE_SIZE));
 
+	/*
+	 * If the channel has been rescinded; process device removal.
+	 */
+	if (channel->rescind)
+		hv_process_channel_removal(channel,
+					   channel->offermsg.child_relid);
 out:
 	tasklet_enable(tasklet);
 
-- 
2.17.1

